File: src/main/java/net/moddingplayground/onastick/api/item/OnAStickItems.java
package net.moddingplayground.onastick.api.item;

import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.minecraft.item.Item;
import net.minecraft.screen.ScreenHandlerFactory;
import net.minecraft.tag.TagKey;
import net.minecraft.text.Text;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import net.moddingplayground.onastick.api.OnAStick;
import net.moddingplayground.onastick.api.screen.LocalAnvilScreenHandler;
import net.moddingplayground.onastick.api.screen.LocalCartographyTableScreenHandler;
import net.moddingplayground.onastick.api.screen.LocalCraftingScreenHandler;
import net.moddingplayground.onastick.api.screen.LocalGrindstoneScreenHandler;
import net.moddingplayground.onastick.api.screen.LocalLoomScreenHandler;
import net.moddingplayground.onastick.api.screen.LocalScreenHandlerFactory;
import net.moddingplayground.onastick.api.screen.LocalSmithingScreenHandler;
import net.moddingplayground.onastick.api.screen.LocalStonecutterScreenHandler;
import net.moddingplayground.onastick.api.tag.OnAStickItemTags;
import net.moddingplayground.onastick.mixin.AnvilBlockAccessor;
import net.moddingplayground.onastick.mixin.CartographyTableBlockAccessor;
import net.moddingplayground.onastick.mixin.CraftingTableBlockAccessor;
import net.moddingplayground.onastick.mixin.GrindstoneBlockAccessor;
import net.moddingplayground.onastick.mixin.LoomBlockAccessor;
import net.moddingplayground.onastick.mixin.SmithingTableBlockAccessor;
import net.moddingplayground.onastick.mixin.StonecutterBlockAccessor;

import java.util.function.Function;

public interface OnAStickItems {
    Item CRAFTING_TABLE_ON_A_STICK = stick("crafting_table_on_a_stick",
        CraftingTableBlockAccessor.getTITLE(), OnAStickItemTags.CRAFTS_CRAFTING_TABLE_ON_A_STICK,
        LocalScreenHandlerFactory.build(LocalCraftingScreenHandler::new, () -> OnAStickItems.CRAFTING_TABLE_ON_A_STICK)
    );

    Item SMITHING_TABLE_ON_A_STICK = stick("smithing_table_on_a_stick",
        SmithingTableBlockAccessor.getSCREEN_TITLE(), OnAStickItemTags.CRAFTS_SMITHING_TABLE_ON_A_STICK,
        LocalScreenHandlerFactory.build(LocalSmithingScreenHandler::new, () -> OnAStickItems.SMITHING_TABLE_ON_A_STICK)
    );

    Item LOOM_ON_A_STICK = stick("loom_on_a_stick",
        LoomBlockAccessor.getTITLE(), OnAStickItemTags.CRAFTS_LOOM_ON_A_STICK,
        LocalScreenHandlerFactory.build(LocalLoomScreenHandler::new, () -> OnAStickItems.LOOM_ON_A_STICK)
    );

    Item CARTOGRAPHY_TABLE_ON_A_STICK = stick("cartography_table_on_a_stick",
        CartographyTableBlockAccessor.getTITLE(), OnAStickItemTags.CRAFTS_CARTOGRAPHY_TABLE_ON_A_STICK,
        LocalScreenHandlerFactory.build(LocalCartographyTableScreenHandler::new, () -> OnAStickItems.CARTOGRAPHY_TABLE_ON_A_STICK)
    );

    Item GRINDSTONE_ON_A_STICK = stick("grindstone_on_a_stick",
        GrindstoneBlockAccessor.getTITLE(), OnAStickItemTags.CRAFTS_GRINDSTONE_ON_A_STICK,
        LocalScreenHandlerFactory.build(LocalGrindstoneScreenHandler::new, () -> OnAStickItems.GRINDSTONE_ON_A_STICK)
    );

    Item ANVIL_ON_A_STICK = stick("anvil_on_a_stick",
        AnvilBlockAccessor.getTITLE(), OnAStickItemTags.CRAFTS_ANVIL_ON_A_STICK,
        LocalScreenHandlerFactory.build(LocalAnvilScreenHandler::new, () -> OnAStickItems.ANVIL_ON_A_STICK)
    );

    Item STONECUTTER_ON_A_STICK = stick("stonecutter_on_a_stick",
        StonecutterBlockAccessor.getTITLE(), OnAStickItemTags.CRAFTS_STONECUTTER_ON_A_STICK,
        LocalScreenHandlerFactory.build(LocalStonecutterScreenHandler::new, () -> OnAStickItems.STONECUTTER_ON_A_STICK)
    );

    private static Item register(String id, Item item) {
        return Registry.register(Registry.ITEM, new Identifier(OnAStick.MOD_ID, id), item);
    }

    private static Item register(String id, Function<Item.Settings, Item> factory) {
        return register(id, factory.apply(new FabricItemSettings().group(OnAStickItemGroups.ALL)));
    }

    private static Item stick(String id, Text title, TagKey<Item> tag, ScreenHandlerFactory screenHandlerFactory) {
        return register(id, settings -> new OnAStickItem(settings.maxCount(1), screenHandlerFactory, title, tag));
    }
}

File: src/main/kotlin/dev/andante/onastick/item/OnAStickItems.kt
package dev.andante.onastick.item

import dev.andante.onastick.OnAStick
import dev.andante.onastick.screen.LocalAnvilScreenHandler
import dev.andante.onastick.screen.LocalCartographyTableScreenHandler
import dev.andante.onastick.screen.LocalCraftingScreenHandler
import dev.andante.onastick.screen.LocalGrindstoneScreenHandler
import dev.andante.onastick.screen.LocalLoomScreenHandler
import dev.andante.onastick.screen.LocalScreenHandlerFactory
import dev.andante.onastick.screen.LocalSmithingScreenHandler
import dev.andante.onastick.screen.LocalStonecutterScreenHandler
import net.minecraft.block.AnvilBlock
import net.minecraft.block.CartographyTableBlock
import net.minecraft.block.CraftingTableBlock
import net.minecraft.block.GrindstoneBlock
import net.minecraft.block.LoomBlock
import net.minecraft.block.SmithingTableBlock
import net.minecraft.block.StonecutterBlock
import net.minecraft.item.Item
import net.minecraft.text.Text
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry


/**
 * Represents On A Stick items.
 */
object OnAStickItems {
    val CRAFTING_TABLE_ON_A_STICK = registerStick("crafting_table_on_a_stick", CraftingTableBlock.TITLE, ::LocalCraftingScreenHandler)
    val SMITHING_TABLE_ON_A_STICK = registerStick("smithing_table_on_a_stick", SmithingTableBlock.SCREEN_TITLE, ::LocalSmithingScreenHandler)
    val LOOM_ON_A_STICK = registerStick("loom_on_a_stick", LoomBlock.TITLE, ::LocalLoomScreenHandler)
    val CARTOGRAPHY_TABLE_ON_A_STICK = registerStick("cartography_table_on_a_stick", CartographyTableBlock.TITLE, ::LocalCartographyTableScreenHandler)
    val GRINDSTONE_ON_A_STICK = registerStick("grindstone_on_a_stick", GrindstoneBlock.TITLE, ::LocalGrindstoneScreenHandler)
    val ANVIL_ON_A_STICK = registerStick("anvil_on_a_stick", AnvilBlock.TITLE, ::LocalAnvilScreenHandler)
    val STONECUTTER_ON_A_STICK = registerStick("stonecutter_on_a_stick", StonecutterBlock.TITLE, ::LocalStonecutterScreenHandler)

    private fun registerStick(id: String, title: Text, localScreenHandlerFactory: LocalScreenHandlerFactory): Item {
        return register(id) { settings -> OnAStickItem(settings.maxCount(1), localScreenHandlerFactory, title) }
    }

    private fun register(id: String, item: Item): Item {
        return Registry.register(Registry.ITEM, Identifier(OnAStick.MOD_ID, id), item)
    }

    private fun register(id: String, item: (Item.Settings) -> Item): Item {
        return register(id, item(Item.Settings().group(OnAStickItemGroups.ALL)))
    }
}
